examples/experimental/dagster-airlift/dagster_airlift/core/load_defs.py

```diff
@@ -152,11 +185,19 @@ class _SerializableDataFromDefs:
 
     # Once over the serialization boundary, these will be combined with an original AssetSpec to create the final AssetSpec.
     # In practice, we attach task-level metadata to the AssetSpecs.
-    spec_transformations: Dict[AssetKey, AssetSpecTransformationData] = {}
+    spec_transformations: Dict[AssetKey, AssetSpecTransformationData]
     # These are the assets which map to tasks within a given dag.
-    keys_in_dag: Dict[str, Set[AssetKey]] = {}
+    keys_in_dag: Dict[str, Set[AssetKey]]
     # This is the asset dependency graph, which we use to determine the leaf assets for a given dag.
-    downstreams_asset_graph: Dict[AssetKey, Set[AssetKey]] = {}
+    downstreams_asset_graph: Dict[AssetKey, Set[AssetKey]]
+    # This is the upstreams asset dependency graph, which we use to determine the topological order of assets.
+    upstreams_asset_graph: Dict[AssetKey, Set[AssetKey]]
+    # The asset keys that correspond to a given task.
+    keys_per_task_handle: Dict[str, Dict[str, Set[AssetKey]]]
+    # Migration state for a given task.
+    migration_state_per_task_handle: Dict[str, Dict[str, Optional[bool]]]
+    # Check keys for a given asset key.
+    check_keys_per_asset_key: Dict[AssetKey, Set[AssetCheckKey]]
```

Comment:
interested in @smackesey 's thoughts here. Seems like we *should* be able to reuse some asset graph primitive here. We are effectively pre-computing an asset graph here so we don't have to redo it all the time.

Reply:
Not following airlift closely so don't have full context here, but I think this is what's going on?

- We're computing a payload from airflow for use as reconstruction metadata
- We're inventing a bespoke structure for it because we don't have a standard structure to serialize defs "state metadata"

If so, yeah there's probably an opportunity to come up with a nice standard abstraction that is similar to `AssetGraph`-- but the only serializable thing I can think of _now_ is a collection of `ExternalAssetNode`, and I'm not sure that would fit here.